### PR TITLE
Composer refactor

### PR DESCRIPTION
**Pull request by du5t**
I was going to clean up my horrible commit history--this is actually a slight U-turn from the edit-posts fork I have going in order to get some sanity around interface elements. (I should have started this one initially...)

This packages up a lot of the static/purefunction components of the Composer into class methods that can then be inherited by the Editor component I introduced to make post editing behave similarly UX-wise and be maintainable without much extra effort. Also, I like packaging monster chunks of JSX into functions with names :)

The post-edit branch I am on will track this one also, so, tell me what you think.

...anyway like I was saying, I was *going* to clean up the horrible commit history (which actually removes a lot of changes!) but squashing these commits with rebase took me to a terrible place. Sorry! :P

### CODE REVIEW

File: ui/com/composer/index.jsx [hunk 1]
```diff
@@ -55,20 +56,19 @@ export default class Composer extends React.Component {
     if (this.props.thread) {
       // thread categorization
       this.isThreadPublic = isThreadPublic(this.props.thread)
-      this.threadChannel = (this.props.thread.value.content.channel || '').trim()
+        this.threadChannel = (this.props.thread.value.content.channel || '').trim()
```
Comment (pfraze) [inline ui/com/composer/index.jsx]: There are some wonky tab situations, like this here. Some of it is alignment, which I'm totally cool with, but can you clean up the stuff like this? Tabs should be at 2 spaces.

File: ui/com/composer/index.jsx [hunk 2]
```diff
@@ -203,6 +203,314 @@ export default class Composer extends React.Component {
...
+  }
+
+  renderAttachBtn(props) {
+    return (props) => {
+      if (!props.isPublic) {
+        if (props.isReply) {
+          return <span/>
+        } else {
+          return <a className="btn disabled">
+          <i className="fa fa-paperclip" /> Attachments not available in PMs</a>
```
Comment (pfraze) [inline ui/com/composer/index.jsx]: Here's another wonky indentation place. I'd prefer this looked like:

```js
return <a className="btn disabled">
  <i className="fa fa-paperclip" /> Attachments not available in PMs
</a>
```

Or was all on one line

File: ui/com/composer/index.jsx [hunk 3]
```diff
@@ -203,6 +203,314 @@ export default class Composer extends React.Component {
...
+      return;
+    
+    this.setState({ isSending: true, isEditing: false })
+
+      // prep text
+      mentionslib.extract(text, (err, mentions) => {
+        if (err) {
+          this.setState({ isSending: false })
+            if (err.conflict)
+            app.issue('Error While Publishing', 'You follow multiple people with the name "'+err.name+'." Resolve this before publishing.')
```
Comment (pfraze) [inline ui/com/composer/index.jsx]: and here too

File: ui/com/composer/index.jsx [hunk 4]
```diff
@@ -203,6 +203,314 @@ export default class Composer extends React.Component {
...
+        const origMsg = this.props.originalMessage
+        const revisionRoot = msg.root ? msg.root :
+                             origMsg.value.content.root ?
+                             origMsg.value.content.root : origMsg.key
+        const revisionMentions = mentions ? mentions : origMsg.value.content.mentions
+        const origKey = origMsg.key
+        
+        // publish edit
+        var edit = schemas.postEdit(text, revisionRoot, revisionMentions, origKey);
+        debugger;
```
Comment (pfraze) [inline ui/com/composer/index.jsx]: leftover debugger statement!

File: ui/com/composer/index.jsx [hunk 5]
```diff
@@ -55,20 +56,19 @@ export default class Composer extends React.Component {
...
+        this.threadRoot = getThreadRoot(this.props.thread)
+        this.threadBranch = threadlib.getLastThreadPost(this.props.thread).key
 
       // extract encryption recipients from thread
       if (Array.isArray(this.props.thread.value.content.recps)) {
         recps = mlib.links(this.props.thread.value.content.recps)
-          .map(function (recp) { return recp.link })
-          .filter(Boolean)
+                    .map(function (recp) { return recp.link })
+                    .filter(Boolean)
```
Comment (pfraze) [inline ui/com/composer/index.jsx]: this I'm cool with :)

**Comment by pfraze** [comment]
Ok other than the formatting stuff, there is some code that has to do with post-edit in here. Did you want that in here? If so then this is really a PR for the post-editing.

Also, that commit history *is* really rough. I think it might be a good idea to create a new branch and apply the diff to it.

File: ui/com/composer/index.jsx (hunk 1 above)
Comment (du5t) [inline ui/com/composer/index.jsx]: whoops! crap, I thought I got all of these. JSX makes a lot of linters and auto-indenters lose their minds ;)

File: ui/com/composer/index.jsx (hunk 4 above)
Comment (du5t) [inline ui/com/composer/index.jsx]: ohhhh nooooo I'm so ashamed... orz

Comment (pfraze) [inline ui/com/composer/index.jsx]: haha no sweat :)

**Comment by du5t** [comment]
sweet, I'll clobber the heck out of these diffs and clean the history up. I'll scale back the refactor and just class-ify the Composer so you have a heads up that an Editor will fit in there once I work the thread rendering out.

**Comment by pfraze** [comment]
sounds good

**Pull request closed without merging**